Sources/PackageCollections/PackageCollections.swift

```diff
@@ -122,14 +122,21 @@ public struct PackageCollections: PackageCollectionsProtocol {
             return callback(.failure(MultipleErrors(errors)))
         }
 
-        // first record the registration
-        self.storage.sources.add(source: source, order: order) { result in
-            switch result {
+        // First try to fetch the collection from the network and store it locally so future operations don't need to access the network
+        self.refreshCollectionFromSource(source: source, order: order) { collectionResult in
+            switch collectionResult {
             case .failure(let error):
                 callback(.failure(error))
-            case .success:
-                // next try to fetch the collection from the network and store it locally so future operations dont need to access the network
-                self.refreshCollectionFromSource(source: source, order: order, callback: callback)
+            case .success(let collection):
+                // next record the registration
+                self.storage.sources.add(source: source, order: order) { sourceResult in
+                    switch sourceResult {
+                    case .failure(let error):
+                        callback(.failure(error))
+                    case .success:
+                        callback(.success(collection))
+                    }
+                }
```

Comment:
the original sequence was intentional for the following use case: say the network is down but I want to add the collection, then we record the addition and when the network is back up a "refresh" will fetch the missing data.

seeing this PR, I can see how it may be less important than validating the collection, so not against this change - just wanted to give background on why it was done this way originally

Reply:
I see. Maybe we can keep this behavior but improve the error messages for `refresh` instead. Right now if a "bad" collection is added, `refresh` would fail but there's no way of knowing which one is the culprit:

```
$ swift-package-collection refresh                                                             
Error: MultipleErrors(errors: [PackageCollections.JSONPackageCollectionProvider.Errors.invalidResponse("Missing Content-Length header")])
```

Perhaps if we include source URL in the message then when user see the error they can decide if they want to remove/keep it.

@neonichu WDYT?

Reply:
I'm not sure presenting the URL would solve the confusion since I can have added and worked with many registries before I do a refresh. Also the discrepancy with `list` would also still be confusing in my mind.

How about we add a flag, e.g. `unverified, to the configuration for collections that have been added while offline and if we see they're actually not valid on next refresh, we remove them, maybe printing a warning about it?

Reply:
have some sort of status for collection makes sense to me, it could help drive other features too

Reply:
I think we can keep track of a `CollectionSource`'s status in the config, though I don't think refresh (or other operations) should remove collections on user's behalf.

Reply:
Config might not be the most appropriate place to store it though since it isn't editable by the user. Putting it in database is overkill IMO.

Reply:
I am also okay with this change as it, the offline mode was a nice to have

Reply:
OK

Reply:
With the "trust confirmation" change, it's easier if we add source instead of fetching it first. Therefore, I am changing the logic here:
1. Add source
2. Fetch source
3. Remove source if fetch fails and it's not because a) we are waiting for user's confirmation on trusting it; b) we have recorded user's preference.